engine/execution/ingestion/stop_at_height.go

```diff
@@ -0,0 +1,79 @@
+package ingestion
+
+import (
+	"fmt"
+	"sync"
+)
+
+type StopAtHeight struct {
+	sync.RWMutex
+	height   uint64
+	crash    bool
+	stopping bool // if stopping process has started, we disallow changes
+	set      bool // whether stop at height has been set at all, it its envisioned most of the time it won't
+}
+
+// NewStopAtHeight creates new empty StopAtHegiht
+func NewStopAtHeight() *StopAtHeight {
+	return &StopAtHeight{
+		stopping: false,
+	}
+}
+
+// Get returns
+// boolean indicating if value is set - for easier comparisons, since its envisions most of the time this struct will be empty
+// height and crash values
+func (s *StopAtHeight) Get() (bool, uint64, bool) {
+	s.RLock()
+	defer s.RUnlock()
+	return s.set, s.height, s.crash
+}
+
+// Try runs function f with current values of height and crash if the values are set
+// f should return true if it started a process of stopping, so no further changes will
+// be accepted.
+// Try returns whatever f returned, or false if f has not been called
+func (s *StopAtHeight) Try(f func(uint64, bool) bool) bool {
+	s.Lock()
+	defer s.Unlock()
+
+	if !s.set {
+		return false
+	}
+
+	started := f(s.height, s.crash)
+
+	if started {
+		s.stopping = true
+	}
```

Comment:
The naming of `started` is very confusing to me. I think it would be much more intuitive, if this was called `commencedStop`
```suggestion
	commencedStop := f(s.height, s.crash)
	if commencedStop {
		s.stopping = true
	}
```